lib/server/monalisa-synchronization/MonALISASynchronizer.js

```diff
@@ -0,0 +1,107 @@
+        this.logger = new Log(MonALISASynchronizer.name);
+        this.monALISAInterface = null;
+    }
+
+    /**
+     * Synchronize with data from MonALISA
+     * Update Data Passes data in local DB
+     * @return {Promise<void>} promise
+     */
+    async synchronizeDataPassesFromMonALISA() {
+        if (!this.monALISAInterface) {
+            this.monALISAInterface = await setupMonALISAInterface();
+        }
+
+        try {
+            this.lastRunNumbers = await this._getLastRunNumbersInformation();
+            const dataPasses = (await this.monALISAInterface.getDataPasses()).filter((dataPass) => this._doesDataPassNeedUpdate(dataPass));
+            // eslint-disable-next-line require-jsdoc
+            const transactionablePipeline = async (dataPass) => {
+                const lhcPeriodDB = await getOrCreateLhcPeriod({ name: extractLhcPeriod(dataPass.name).name });
+                await DataPassRepository.upsert({ lhcPeriodId: lhcPeriodDB.id, ...dataPass });
+                const dataPassDB = await DataPassRepository.findOne({ where: { name: dataPass.name } });
+                const { runNumbers } = await this.monALISAInterface.getDataPassDetails(dataPassDB.description);
+                const runs = await RunRepository.findAll(dataSource.createQueryBuilder().where('runNumber').oneOf(...runNumbers));
+                await dataPassDB.addRuns(runs, { ignoreDuplicates: true });
+            };
+
+            for (const dataPass of dataPasses) {
+                await dataSource.transaction(() => transactionablePipeline(dataPass));
+            }
+            this.logger.info('Synchronization completed');
+        } catch (error) {
+            this.logger.error(error.message);
+        } finally {
+            this.lastRunNumbers = null;
+        }
+    }
+
+    /**
+     * Check wether data pass needs to be updated
+     * @param {DataPass} dataPass data pass
+     * @return {boolean} true if it is needed to update data pass
+     */
+    _doesDataPassNeedUpdate(dataPass) {
+        const isLastRunNumberSame = dataPass.lastRun && dataPass.lastRun === this.lastRunNumbers[dataPass.name];
+        return !isLastRunNumberSame;
+    }
+
+    /**
+     * Fetch last runs information telling whether some data pass must be updated
+     * last run number of a data pass is the greatest run number associated with given data pass
+     * @return {Object<string, number>} mapping data pass name -> its last run
+     */
+    async _getLastRunNumbersInformation() {
+        const queryBuilder = dataSource.createQueryBuilder()
+            .set('raw', true)
+            .set('attributes', ['name'])
+            .include((sequelize) => ({
+                association: 'runs',
+                attributes: [[sequelize.fn('max', sequelize.col('runs.run_number')), 'lastRun']],
+                through: { attributes: [] },
+            }))
+            .groupBy('name');
+
+        const lastRunsData = await DataPassRepository.findAll(queryBuilder);
```

Comment:
It's more `dataPasses` than `lastRunData`

Reply:
applied